<a href="https://colab.research.google.com/github/markmaloba/W4-independent-project/blob/master/Moringa_Data_Science_Prep_W4_Independent_Project_2019_10_Mark_Maloba_Python_notebook_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
#Import libraries

import numpy as np
import pandas as pd


In [0]:
#Load datasets
AUTOdf = pd.read_csv('/content/Autolib_dataset.csv')
AUTOdf.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

In [0]:
ALTdf = pd.read_csv('/content/Autolib_dataset.csv')
ALTdf.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

# Cleaning

In [0]:
#The future stations do not help with our analysis of current events, so the related "Scheduled at" column, by extension, isn't necessary. 
#That and the comments column also happen to be the 2most incomplete columns as seen below:
AUTOdf.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
Address                5000 non-null object
Cars                   5000 non-null int64
Bluecar counter        5000 non-null int64
Utilib counter         5000 non-null int64
Utilib 1.4 counter     5000 non-null int64
Charge Slots           5000 non-null int64
Charging Status        5000 non-null object
City                   5000 non-null object
Displayed comment      111 non-null object
ID                     5000 non-null object
Kind                   5000 non-null object
Geo point              5000 non-null object
Postal code            5000 non-null int64
Public name            5000 non-null object
Rental status          5000 non-null object
Scheduled at           47 non-null object
Slots                  5000 non-null int64
Station type           5000 non-null object
Status                 5000 non-null object
Subscription status    5000 non-null object
year                   

In [0]:
#Hence we can drop them
#AUTOdf.drop(["Displayed comment","Scheduled at"], axis = 1, inplace= True)
AUTOdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 23 columns):
Address                5000 non-null object
Cars                   5000 non-null int64
Bluecar counter        5000 non-null int64
Utilib counter         5000 non-null int64
Utilib 1.4 counter     5000 non-null int64
Charge Slots           5000 non-null int64
Charging Status        5000 non-null object
City                   5000 non-null object
ID                     5000 non-null object
Kind                   5000 non-null object
Geo point              5000 non-null object
Postal code            5000 non-null int64
Public name            5000 non-null object
Rental status          5000 non-null object
Slots                  5000 non-null int64
Station type           5000 non-null object
Status                 5000 non-null object
Subscription status    5000 non-null object
year                   5000 non-null int64
month                  5000 non-null int64
day                   

In [0]:
#Computing IQR for each column to detect Outliers

Q1 = AUTOdf.quantile(0.25)
Q3 = AUTOdf.quantile(0.75)
IQR = Q3 - Q1

print(IQR)

Cars                      4.0
Bluecar counter           4.0
Utilib counter            0.0
Utilib 1.4 counter        0.0
Charge Slots              0.0
Postal code           17308.0
Slots                     3.0
year                      0.0
month                     0.0
day                       4.0
hour                     12.0
minute                   30.0
dtype: float64


In [0]:
#Removing the outliers

print((AUTOdf  < (Q1 - 1.5 * IQR)) |(AUTOdf > (Q3 + 1.5 * IQR)))

AUTOdf2 = AUTOdf[~((AUTOdf < (Q1 - 1.5 * IQR)) |(AUTOdf > (Q3 + 1.5 * IQR))).any(axis=1)]
AUTOdf2.shape

      Address  Bluecar counter   Cars  ...  minute  month   year
0       False            False  False  ...   False  False  False
1       False            False  False  ...   False  False  False
2       False            False  False  ...   False  False  False
3       False            False  False  ...   False  False  False
4       False            False  False  ...   False  False  False
...       ...              ...    ...  ...     ...    ...    ...
4995    False            False  False  ...   False  False  False
4996    False            False  False  ...   False  False  False
4997    False            False  False  ...   False  False  False
4998    False            False  False  ...   False  False  False
4999    False            False  False  ...   False  False  False

[5000 rows x 23 columns]


(3317, 23)

In [0]:
#Combining Date & Time in preparation for Analysis
#We will also drop the unnecessary split date & time columns

# AUTOdf2['Date'] = AUTOdf2["year"].map(str) +':'+ AUTOdf2["month"].map(str) +':'+ AUTOdf2["day"].map(str)
# AUTOdf2.head()
# AUTOdf2['Time']=AUTOdf2["hour"].map(str) +':'+ AUTOdf2["minute"].map(str)
# AUTOdf2.head()
# AUTOdf2['Date.Time']= AUTOdf2["Date"].map(str) +':'+ AUTOdf2["Time"].map(str)
# AUTOdf2.head()
AUTOdf2.drop(['year','month','day','minute','Date','Time'],axis=1,inplace=True)
AUTOdf2.head()

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4117: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Address  Cars  ...  hour       Date.Time
0                 2 Avenue de Suffren     0  ...    11  2018:4:8:11:43
1           145 Rue Raymond Losserand     6  ...     7   2018:4:6:7:24
4  6 avenue de la Porte de Champerret     3  ...    17  2018:4:8:17:23
5                8 Boulevard Voltaire     0  ...     7    2018:4:6:7:2
6                      37 rue Leblanc     0  ...    18  2018:4:8:18:20

[5 rows x 20 columns]

In [0]:
#Last cleaning action is to check for & deal with duplicate values (if any) in our DataFrame

AUTOdf2.duplicated().sum()

0

# Analysis

In [0]:
#Subsetting: Separate Bluecar activity for Paris in the month of April
AUTOdf3 = AUTOdf2[(AUTOdf2['City']=='Paris') & (AUTOdf2['Bluecar counter']!=0)]
AUTOdf3.head(5)

Address  Cars  ...  hour       Date.Time
1            145 Rue Raymond Losserand     6  ...     7   2018:4:6:7:24
4   6 avenue de la Porte de Champerret     3  ...    17  2018:4:8:17:23
8                34 avenue Jean Moulin     1  ...    22  2018:4:2:22:58
9         41 boulevard de Rochechouart     6  ...    15   2018:4:4:15:2
17              182 Boulevard Voltaire     6  ...    12  2018:4:9:12:14

[5 rows x 20 columns]

In [0]:
#Grouping the data by station

AUTOdf4 = AUTOdf3.sort_values(['Date.Time']).groupby(['ID','Date.Time']).sum()
AUTOdf4.head(15)

Cars  Bluecar counter  ...  Slots  hour
ID                   Date.Time                              ...             
paris-25aout1944-8   2018:4:1:18:40     1                1  ...      2    18
                     2018:4:2:12:49     2                2  ...      1    12
                     2018:4:2:7:37      4                4  ...      0     7
                     2018:4:4:18:48     1                1  ...      1    18
                     2018:4:4:3:10      3                3  ...      0     3
paris-adolphemax-6   2018:4:2:11:1      4                4  ...      0    11
paris-adolphemille-8 2018:4:1:1:53      1                1  ...      1     1
                     2018:4:3:12:25     2                2  ...      1    12
                     2018:4:3:21:53     4                4  ...      0    21
paris-adolpheyvon-6  2018:4:2:14:15     5                5  ...      0    14
                     2018:4:2:19:14     5                5  ...      1    19
                     2018:4:4:23:18     5                5  ...      1    23
                     2018:4:7:23:11     4                4  ...      2    23
paris-alesia-33      2018:4:7:7:43      5                5  ...      1     7
paris-alexandrie-7   2018:4:5:16:15     3                3  ...      1    16

[15 rows x 8 columns]

In [0]:
#Custom function to indicate car pickup vs return. +ve = pickup, -ve = return 

AUTOdf4['rt_or_pick'] = AUTOdf4.groupby(['ID'])['Bluecar counter'].apply(lambda x:  x.shift(1) - x)
AUTOdf4['rt_or_pick']
AUTOdf4.head(15)

Cars  Bluecar counter  ...  hour  rt_or_pick
ID                   Date.Time                              ...                  
paris-25aout1944-8   2018:4:1:18:40     1                1  ...    18         NaN
                     2018:4:2:12:49     2                2  ...    12        -1.0
                     2018:4:2:7:37      4                4  ...     7        -2.0
                     2018:4:4:18:48     1                1  ...    18         3.0
                     2018:4:4:3:10      3                3  ...     3        -2.0
paris-adolphemax-6   2018:4:2:11:1      4                4  ...    11         NaN
paris-adolphemille-8 2018:4:1:1:53      1                1  ...     1         NaN
                     2018:4:3:12:25     2                2  ...    12        -1.0
                     2018:4:3:21:53     4                4  ...    21        -2.0
paris-adolpheyvon-6  2018:4:2:14:15     5                5  ...    14         NaN
                     2018:4:2:19:14     5                5  ...    19         0.0
                     2018:4:4:23:18     5                5  ...    23         0.0
                     2018:4:7:23:11     4                4  ...    23         1.0
paris-alesia-33      2018:4:7:7:43      5                5  ...     7         NaN
paris-alexandrie-7   2018:4:5:16:15     3                3  ...    16         NaN

[15 rows x 9 columns]

#Answers to IP Questions

In [0]:
#most popular hour of the day for picking up a shared electric car (Bluecar) 
#in the city of Paris over the month of April 2018

ALTdf = ALTdf.rename(columns={'Bluecar counter': 'blue_count'})
Blu_Paris = ALTdf[(ALTdf.City == 'Paris') & (ALTdf.blue_count == 0)].groupby(['hour'])[blue_count].sum()
Blu_Paris.sort_values(ascending = False).head(1)

NameError: ignored

In [0]:
#Most popular hour for returning cars

AUTOdf4[(AUTOdf4['rt_or_pick'] < 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

hour
23   -5.0
Name: rt_or_pick, dtype: float64

Q1 ans = 2300H


In [0]:
#Most popular hour for picking cars

AUTOdf4[(AUTOdf4['rt_or_pick'] > 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

hour
21    31.0
Name: rt_or_pick, dtype: float64

Ans: 2100H


In [0]:
#Overall most popular picking station

AUTOdf4['rt_or_pick']=AUTOdf4['rt_or_pick'].abs()
AUTOdf4.head()
AUTOdf4.groupby('ID')['rt_or_pick'].sum().sort_values(ascending=False).head(1)

ID
paris-raymondpitet-7    11.0
Name: rt_or_pick, dtype: float64

Ans: paris-raymondpitet-7

In [0]:
#Overall most popular picking station at the most popular picking hour?

AUTOdf4[(AUTOdf4['rt_or_pick'] > 0) & (AUTOdf4['hour']==21)].groupby(['hour','ID'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

hour  ID                 
21    paris-saintmande-94    4.0
Name: rt_or_pick, dtype: float64

Ans: paris-saintmande-94

In [0]:
#Most popular postcode for picking blue cars
AUTOdf4[(AUTOdf4['rt_or_pick'] > 0)].groupby(['ID','Postal code'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

ID                    Postal code
paris-raymondpitet-7  75017          11.0
Name: rt_or_pick, dtype: float64

Ans: 75017, and yes the overall most popular station belongs to that postcode          

# Results for the above, considering Utilib

In [0]:
AUTOdf5 = AUTOdf4
AUTOdf5.head(5)


Cars  Bluecar counter  ...  hour  rt_or_pick
ID                 Date.Time                              ...                  
paris-25aout1944-8 2018:4:1:18:40     1                1  ...    18         NaN
                   2018:4:2:12:49     2                2  ...    12         1.0
                   2018:4:2:7:37      4                4  ...     7         2.0
                   2018:4:4:18:48     1                1  ...    18         3.0
                   2018:4:4:3:10      3                3  ...     3         2.0

[5 rows x 9 columns]

In [0]:
AUTOdf5['rt_or_pick'] = AUTOdf5.groupby(['ID'])['Utilib counter'].apply(lambda x:  x.shift(1) - x)
AUTOdf5['rt_or_pick']
AUTOdf5.head(15)

Cars  Bluecar counter  ...  hour  rt_or_pick
ID                   Date.Time                              ...                  
paris-25aout1944-8   2018:4:1:18:40     1                1  ...    18         NaN
                     2018:4:2:12:49     2                2  ...    12         0.0
                     2018:4:2:7:37      4                4  ...     7         0.0
                     2018:4:4:18:48     1                1  ...    18         0.0
                     2018:4:4:3:10      3                3  ...     3         0.0
paris-adolphemax-6   2018:4:2:11:1      4                4  ...    11         NaN
paris-adolphemille-8 2018:4:1:1:53      1                1  ...     1         NaN
                     2018:4:3:12:25     2                2  ...    12         0.0
                     2018:4:3:21:53     4                4  ...    21         0.0
paris-adolpheyvon-6  2018:4:2:14:15     5                5  ...    14         NaN
                     2018:4:2:19:14     5                5  ...    19         0.0
                     2018:4:4:23:18     5                5  ...    23         0.0
                     2018:4:7:23:11     4                4  ...    23         0.0
paris-alesia-33      2018:4:7:7:43      5                5  ...     7         NaN
paris-alexandrie-7   2018:4:5:16:15     3                3  ...    16         NaN

[15 rows x 9 columns]

In [0]:
#Most popular hour for returning cars
AUTOdf5[(AUTOdf5['rt_or_pick'] < 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

Series([], Name: rt_or_pick, dtype: float64)

In [0]:
#Most popular hour for picking cars

AUTOdf5[(AUTOdf5['rt_or_pick'] > 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

Series([], Name: rt_or_pick, dtype: float64)

In [0]:
#Overall most popular picking station

AUTOdf5['rt_or_pick']=AUTOdf5['rt_or_pick'].abs()
AUTOdf5.head()
AUTOdf5.groupby('ID')['rt_or_pick'].sum().sort_values(ascending=False).head(1)

ID
paris-wagram-77    0.0
Name: rt_or_pick, dtype: float64

**Ans: The most popular station has now changed to paris-wagram-77**




In [0]:
#Most popular postcode for picking Utilib cars
AUTOdf5[(AUTOdf5['rt_or_pick'] > 0)].groupby(['ID','Postal code'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

Series([], Name: rt_or_pick, dtype: float64)

# Results for the above, considering Utilib 1.4

In [0]:
AUTOdf6 = AUTOdf4
AUTOdf6.head(5)

Cars  Bluecar counter  ...  hour  rt_or_pick
ID                 Date.Time                              ...                  
paris-25aout1944-8 2018:4:1:18:40     1                1  ...    18         NaN
                   2018:4:2:12:49     2                2  ...    12         0.0
                   2018:4:2:7:37      4                4  ...     7         0.0
                   2018:4:4:18:48     1                1  ...    18         0.0
                   2018:4:4:3:10      3                3  ...     3         0.0

[5 rows x 9 columns]

In [0]:
AUTOdf6['rt_or_pick'] = AUTOdf6.groupby(['ID'])['Utilib 1.4 counter'].apply(lambda x:  x.shift(1) - x)
AUTOdf6['rt_or_pick']
AUTOdf6.head(15)

Cars  Bluecar counter  ...  hour  rt_or_pick
ID                   Date.Time                              ...                  
paris-25aout1944-8   2018:4:1:18:40     1                1  ...    18         NaN
                     2018:4:2:12:49     2                2  ...    12         0.0
                     2018:4:2:7:37      4                4  ...     7         0.0
                     2018:4:4:18:48     1                1  ...    18         0.0
                     2018:4:4:3:10      3                3  ...     3         0.0
paris-adolphemax-6   2018:4:2:11:1      4                4  ...    11         NaN
paris-adolphemille-8 2018:4:1:1:53      1                1  ...     1         NaN
                     2018:4:3:12:25     2                2  ...    12         0.0
                     2018:4:3:21:53     4                4  ...    21         0.0
paris-adolpheyvon-6  2018:4:2:14:15     5                5  ...    14         NaN
                     2018:4:2:19:14     5                5  ...    19         0.0
                     2018:4:4:23:18     5                5  ...    23         0.0
                     2018:4:7:23:11     4                4  ...    23         0.0
paris-alesia-33      2018:4:7:7:43      5                5  ...     7         NaN
paris-alexandrie-7   2018:4:5:16:15     3                3  ...    16         NaN

[15 rows x 9 columns]

In [0]:
#Most popular hour for returning cars
AUTOdf6[(AUTOdf6['rt_or_pick'] < 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

Series([], Name: rt_or_pick, dtype: float64)

In [0]:
#Most popular hour for picking cars
AUTOdf6[(AUTOdf6['rt_or_pick'] > 0)].groupby(['hour'])['rt_or_pick'].sum().sort_values(ascending = False).head(1)

Series([], Name: rt_or_pick, dtype: float64)

In [0]:
#Overall most popular picking station

AUTOdf6['rt_or_pick']=AUTOdf6['rt_or_pick'].abs()
AUTOdf6.head()
AUTOdf6.groupby('ID')['rt_or_pick'].sum().sort_values(ascending=False).head(1)

ID
paris-wagram-77    0.0
Name: rt_or_pick, dtype: float64

**Ans: Utilib and Utilib 1.4 have paris-wagram-77 as the most popular station**